### DISCUSSION TITLE
Why does `offsetForArgs` add 1 to `after`?

### DISCUSSION THREAD

**Discussion by danieljvdm** [hayes/pothos#943] [General]
https://github.com/hayes/pothos/blob/main/packages/plugin-relay/src/utils/connections.ts#L62

I don't understand why 1 is added here. If I have a B64 string that converts to `After:20`, I want my SQL query to be `OFFSET 20`, not `OFFSET 21`, otherwise I'm skipping a row.

GraphQL Relay JS doesn't do this: https://github.com/graphql/graphql-relay-js/blob/main/src/connection/arrayConnection.ts#L136C1-L141C2

What am I missing here?

**Comment by hayes** [comment]
Thanks for the question!

Basically the way this works is if you have a response like:

```js
{
  someConnection: {
    pageInfo: {
      hasPreviousPage: false,
      hasNextPage: true,
      startCursor: 'offset:0',
      endCursor: 'offset:2',
    },
    edges: [
      {
        cursor: 'offset:0',
        node: {
          id: '1',
        },
      },
      {
        cursor: 'offset:1',
        node: {
          id: '2',
        },
      },
      {
        cursor: 'offset:2',
        node: {
          id: '3',
        },
      },
    ],
  },
};
```

The query for the next page would look something like:
```graphql
query {
  someConnection(after: "offset:2", first: 3) {
    ...
  }
}
```

Based on your question, I think the important thing to remember is the `before` and `after` take cursors, and a cursor points to a specific node.  So the cursor the the last item shouldn't be something like `After:20`, it should be `offset:19`, or `id:20` or however you want to represent it, but the goal is to point to the location of the item, not the item that follows it.

**Reply by hayes** [reply]
FWIW, graphql-relay-js adds the + 1 to the offset here: https://github.com/graphql/graphql-relay-js/blob/main/src/connection/arrayConnection.ts#L63

**Reply by danieljvdm** [reply]
Got it, this is helpful. I guess it was just a classic "array indexes start at 0" problem and the cursors being used here are just offsets and I was running into an off-by-one problem. This explanation makes sense, thanks a bunch!
